notes/overriding.md

```diff
@@ -0,0 +1,942 @@
+
+### Implementation
+
+Recomputing the fixpoint at each field access is wasteful, because recursive
+records as functions satisfy the following property: as long as it is not
+merged, the `self` argument is constant. Thus, we can memoize the fixpoint and
+only invalidate `self` on a merge. Also, fields that do not depend on a
+recursive variable can be hoisted out of the function, using a representation
+like:
+
+```
+repr(r) := {
+  a = 1;
+  b = fun self => self.a + 1;
+}
+```
+
+Concretely, each field of a record may either be a thunk as usual, in the case
+of non recursive expressions like `a`, or a thunk together with the original
+expression in the recursive case. All record operations excepted merge operate
+on the thunk. A merge operation, on the other hand, restores the original
+expression again.
+
+There's more potential optimizations, but this first step should be a reasonable
+trade-off between implementation complexity and performance. See
+[#103](https://github.com/tweag/nickel/issues/103) for more details.
+
+### Scoping
+
+Should a record be able to access a yet undefined field because it is expected
+to be provided by a subsequent merge? Two possible approaches:
+
+ - **dynamic scoping**: records can reference fields that are not explicitly
+   defined locally, such as:
+   ```nickel
+   {a = b} & {b = 1}
+   ```
+   Dynamic scoping have a number of issues, and is usually considered bad
+   practice.
+ - **lexical scoping**: as currently, require self-referenced fields to be
+   defined locally. Note that thanks to contracts, one can require the existence
+   of a field without defining it. For example, we could write the previous
+   example as:
+   ```nickel
+   {a = b, b | Num} & {b = 1}
+   ```
+   This is also a better practice to explicitly state the fields whose presence
+   is assumed in general.
+
+This RFC proposes to adopt **lexical scoping**. We could have an even lighter
+syntax, such as `{a = b, b} & {b = 1}` for requiring the presence of a field
+`b`.
+
+### Priorities
+
+#### Levels
+
+This RFC proposes to adopt the following ordered set of priorities:
+
+- `default` is the bottom element
+- Integers priorities in the middle
+- `force` is the top element
+
+That is, `Priorities := default \/ {n | n integer} \/ top` with `default <= ...
+<= -1 <= 0 <= 1 <= ... <= force`.
+
+If not specified, the normal priority (the default priority, no to be confused
+with the `default` priority) is `0`. This provides an infinite supply of
+priorities both below (`default \/ {n | n < 0}`) as well as above (`force \/ {n
+| n > 0}`).
+
+Integer priorities are specified using the `priority` keyword. Defining more
+than one priority in the same meta-value is an error.
+
+Example:
+```nickel
+{
+  foo | Num
+      | default = 1,
+
+  bar | Str,
+  //equivalent to `bar | Str | priority 0`
+
+  baz.boo.bor | priority -4 = "value",
+
+  final | force = `CantOverrideMe,
+}
+```
+
+#### Recursive priorities
```

Comment:
I think that this section will need refinement. Its not as well understood as it ought to.